File: scripts/scan/manifest.py
"""F-036 — deterministic tech_stack and project-name extraction.

Reads package.json / pyproject.toml / Cargo.toml / go.mod when present and
returns a normalized dict keyed by ``runtime`` / ``language`` / ``test`` /
``build`` / ``min_version``.

Pure functions, no I/O outside of reading the listed manifest files.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Optional

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Python 3.10 backport


_NODE_TEST_PRIORITY = ("vitest", "jest", "mocha", "ava", "node:test")


def extract_tech_stack(root: Path) -> dict:
    """Return a tech_stack dict for the most prominent manifest under ``root``.

    Detection order: package.json > pyproject.toml > Cargo.toml > go.mod.
    Returns an empty dict when no recognized manifest is present.
    """
    root = Path(root)
    if (root / "package.json").is_file():
        return _detect_node(root)
    if (root / "pyproject.toml").is_file():
        return _detect_python(root)
    if (root / "Cargo.toml").is_file():
        return _detect_rust(root)
    if (root / "go.mod").is_file():
        return _detect_go(root)
    return {}


def extract_project_name(root: Path) -> Optional[str]:
    """Return the declared project name, falling back to the directory basename.

    Priority: package.json:.name > pyproject.toml:[project].name >
    Cargo.toml:[package].name > go.mod module path tail > directory basename.
    Returns ``None`` only if the directory does not exist.
    """
    root = Path(root)
    if not root.exists():
        return None

    pkg = root / "package.json"
    if pkg.is_file():
        try:
            data = json.loads(pkg.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            data = {}
        if isinstance(data.get("name"), str) and data["name"]:
            return data["name"]

    pyproj = root / "pyproject.toml"
    if pyproj.is_file():
        try:
            data = tomllib.loads(pyproj.read_text(encoding="utf-8"))
        except tomllib.TOMLDecodeError:
            data = {}
        name = data.get("project", {}).get("name")
        if isinstance(name, str) and name:
            return name

    cargo = root / "Cargo.toml"
    if cargo.is_file():
        try:
            data = tomllib.loads(cargo.read_text(encoding="utf-8"))
        except tomllib.TOMLDecodeError:
            data = {}
        name = data.get("package", {}).get("name")
        if isinstance(name, str) and name:
            return name

    gomod = root / "go.mod"
    if gomod.is_file():
        match = re.search(r"^module\s+(\S+)", gomod.read_text(encoding="utf-8"), re.MULTILINE)
        if match:
            return match.group(1).rsplit("/", 1)[-1]

    return root.name


def _detect_node(root: Path) -> dict:
    try:
        data = json.loads((root / "package.json").read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return {}

    deps = {**data.get("dependencies", {}), **data.get("devDependencies", {})}

    has_typescript = "typescript" in deps or (root / "tsconfig.json").is_file()
    language = "typescript" if has_typescript else "javascript"

    test_runner = next((name for name in _NODE_TEST_PRIORITY if name in deps), "")

    build = ""
    if "vite" in deps or any((root / f).is_file() for f in ("vite.config.ts", "vite.config.js")):
        build = "vite"
    elif "webpack" in deps or any((root / f).is_file() for f in ("webpack.config.ts", "webpack.config.js")):
        build = "webpack"
    elif "esbuild" in deps:
        build = "esbuild"

    engines_node = data.get("engines", {}).get("node", "")
    min_version = engines_node.lstrip(">=^~ ") if isinstance(engines_node, str) else ""

    stack = {
        "runtime": "node",
        "language": language,
    }
    if test_runner:
        stack["test"] = test_runner
    if build:
        stack["build"] = build
    if min_version:
        stack["min_version"] = min_version
    return stack


def _detect_python(root: Path) -> dict:
    try:
        data = tomllib.loads((root / "pyproject.toml").read_text(encoding="utf-8"))
    except tomllib.TOMLDecodeError:
        return {}

    project = data.get("project", {})
    deps_blob: list[str] = list(project.get("dependencies", []))
    optional = project.get("optional-dependencies", {})
    if isinstance(optional, dict):
        for group in optional.values():
            if isinstance(group, list):
                deps_blob.extend(group)
    extras = data.get("dependency-groups", {})
    if isinstance(extras, dict):
        for group in extras.values():
            if isinstance(group, list):
                deps_blob.extend(group)

    has_pytest = any("pytest" in dep for dep in deps_blob) or "pytest" in data.get("tool", {})
    test_runner = "pytest" if has_pytest else ""

    build_backend = data.get("build-system", {}).get("build-backend", "")
    build = ""
    if "setuptools" in build_backend:
        build = "setuptools"
    elif "poetry" in build_backend:
        build = "poetry"
    elif "hatchling" in build_backend:
        build = "hatch"
    elif "flit" in build_backend:
        build = "flit"

    requires_python = project.get("requires-python", "")
    min_version = requires_python.lstrip(">=^~ ") if isinstance(requires_python, str) else ""

    stack = {
        "runtime": "python",
        "language": "python",
    }
    if test_runner:
        stack["test"] = test_runner
    if build:
        stack["build"] = build
    if min_version:
        stack["min_version"] = min_version
    return stack


def _detect_rust(root: Path) -> dict:
    try:
        data = tomllib.loads((root / "Cargo.toml").read_text(encoding="utf-8"))
    except tomllib.TOMLDecodeError:
        return {}

    package = data.get("package", {})
    rust_version = package.get("rust-version", "")

    stack = {
        "runtime": "rust",
        "language": "rust",
        "test": "cargo",
        "build": "cargo",
    }
    if isinstance(rust_version, str) and rust_version:
        stack["min_version"] = rust_version
    return stack


def _detect_go(root: Path) -> dict:
    text = (root / "go.mod").read_text(encoding="utf-8")
    match = re.search(r"^go\s+(\S+)", text, re.MULTILINE)
    min_version = match.group(1) if match else ""

    stack = {
        "runtime": "go",
        "language": "go",
        "test": "go",
        "build": "go",
    }
    if min_version:
        stack["min_version"] = min_version
    return stack

File: src/scan/manifest.ts
/**
 * Tech-stack + project-name extraction across language manifests
 * (F-101 port of `scripts/scan/manifest.py`, F-036).
 *
 * Detection priority — first match wins:
 *
 *   package.json → pyproject.toml → Cargo.toml → go.mod
 *
 * Pure: only reads the four listed manifest files. TOML parsing uses
 * smol-toml (already a runtime dep) so pyproject / Cargo manifests
 * are first-class.
 *
 * @module scan/manifest
 */

import {readFileSync, statSync} from 'node:fs';
import {basename, join} from 'node:path';
import {parse as tomlParse} from 'smol-toml';

const NODE_TEST_PRIORITY: ReadonlyArray<string> = ['vitest', 'jest', 'mocha', 'ava', 'node:test'];

/** Output shape of {@link extractTechStack}; keys are optional per Python parity. */
export interface TechStack {
  runtime?: string;
  language?: string;
  test?: string;
  build?: string;
  min_version?: string;
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

/** Returns the tech_stack dict for the most prominent manifest. */
export function extractTechStack(root: string): TechStack {
  if (isFile(join(root, 'package.json'))) {
    return detectNode(root);
  }
  if (isFile(join(root, 'pyproject.toml'))) {
    return detectPython(root);
  }
  if (isFile(join(root, 'Cargo.toml'))) {
    return detectRust(root);
  }
  if (isFile(join(root, 'go.mod'))) {
    return detectGo(root);
  }
  return {};
}

/** Returns the declared project name, falling back to the directory basename. */
export function extractProjectName(root: string): string | null {
  try {
    statSync(root);
  } catch {
    return null;
  }

  const pkgPath = join(root, 'package.json');
  if (isFile(pkgPath)) {
    try {
      const data: unknown = JSON.parse(readFileSync(pkgPath, 'utf-8'));
      if (isPlainObject(data) && typeof data['name'] === 'string' && data['name']) {
        return data['name'];
      }
    } catch {
      // ignore
    }
  }

  const pyproj = join(root, 'pyproject.toml');
  if (isFile(pyproj)) {
    try {
      const data = tomlParse(readFileSync(pyproj, 'utf-8'));
      const project = isPlainObject(data['project']) ? (data['project'] as Record<string, unknown>) : {};
      if (typeof project['name'] === 'string' && project['name']) {
        return project['name'];
      }
    } catch {
      // ignore
    }
  }

  const cargo = join(root, 'Cargo.toml');
  if (isFile(cargo)) {
    try {
      const data = tomlParse(readFileSync(cargo, 'utf-8'));
      const pkg = isPlainObject(data['package']) ? (data['package'] as Record<string, unknown>) : {};
      if (typeof pkg['name'] === 'string' && pkg['name']) {
        return pkg['name'];
      }
    } catch {
      // ignore
    }
  }

  const gomod = join(root, 'go.mod');
  if (isFile(gomod)) {
    try {
      const text = readFileSync(gomod, 'utf-8');
      const match = /^module\s+(\S+)/m.exec(text);
      if (match !== null) {
        return match[1]!.split('/').pop() ?? basename(root);
      }
    } catch {
      // ignore
    }
  }

  return basename(root);
}

function detectNode(root: string): TechStack {
  let data: unknown;
  try {
    data = JSON.parse(readFileSync(join(root, 'package.json'), 'utf-8'));
  } catch {
    return {};
  }
  if (!isPlainObject(data)) {
    return {};
  }

  const dependencies = isPlainObject(data['dependencies']) ? (data['dependencies'] as Record<string, unknown>) : {};
  const devDependencies = isPlainObject(data['devDependencies']) ? (data['devDependencies'] as Record<string, unknown>) : {};
  const deps: Record<string, unknown> = {...dependencies, ...devDependencies};

  const hasTypescript = 'typescript' in deps || isFile(join(root, 'tsconfig.json'));
  const language = hasTypescript ? 'typescript' : 'javascript';

  const testRunner = NODE_TEST_PRIORITY.find((name) => name in deps) ?? '';

  let build = '';
  if (
    'vite' in deps ||
    isFile(join(root, 'vite.config.ts')) ||
    isFile(join(root, 'vite.config.js'))
  ) {
    build = 'vite';
  } else if (
    'webpack' in deps ||
    isFile(join(root, 'webpack.config.ts')) ||
    isFile(join(root, 'webpack.config.js'))
  ) {
    build = 'webpack';
  } else if ('esbuild' in deps) {
    build = 'esbuild';
  }

  const engines = isPlainObject(data['engines']) ? (data['engines'] as Record<string, unknown>) : {};
  const enginesNode = typeof engines['node'] === 'string' ? engines['node'] : '';
  const minVersion = enginesNode.replace(/^[\s>=^~]+/, '');

  const stack: TechStack = {runtime: 'node', language};
  if (testRunner) {
    stack.test = testRunner;
  }
  if (build) {
    stack.build = build;
  }
  if (minVersion) {
    stack.min_version = minVersion;
  }
  return stack;
}

function detectPython(root: string): TechStack {
  let data: unknown;
  try {
    data = tomlParse(readFileSync(join(root, 'pyproject.toml'), 'utf-8'));
  } catch {
    return {};
  }
  if (!isPlainObject(data)) {
    return {};
  }
  const project = isPlainObject(data['project']) ? (data['project'] as Record<string, unknown>) : {};
  const depsBlob: string[] = [];
  if (Array.isArray(project['dependencies'])) {
    for (const d of project['dependencies'] as unknown[]) {
      if (typeof d === 'string') {
        depsBlob.push(d);
      }
    }
  }
  const optional = project['optional-dependencies'];
  if (isPlainObject(optional)) {
    for (const group of Object.values(optional)) {
      if (Array.isArray(group)) {
        for (const d of group) {
          if (typeof d === 'string') {
            depsBlob.push(d);
          }
        }
      }
    }
  }
  const extras = data['dependency-groups'];
  if (isPlainObject(extras)) {
    for (const group of Object.values(extras)) {
      if (Array.isArray(group)) {
        for (const d of group) {
          if (typeof d === 'string') {
            depsBlob.push(d);
          }
        }
      }
    }
  }
  const tool = isPlainObject(data['tool']) ? (data['tool'] as Record<string, unknown>) : {};
  const hasPytest = depsBlob.some((dep) => dep.includes('pytest')) || 'pytest' in tool;
  const testRunner = hasPytest ? 'pytest' : '';

  const buildSystem = isPlainObject(data['build-system']) ? (data['build-system'] as Record<string, unknown>) : {};
  const buildBackend = typeof buildSystem['build-backend'] === 'string' ? buildSystem['build-backend'] : '';
  let build = '';
  if (buildBackend.includes('setuptools')) {
    build = 'setuptools';
  } else if (buildBackend.includes('poetry')) {
    build = 'poetry';
  } else if (buildBackend.includes('hatchling')) {
    build = 'hatch';
  } else if (buildBackend.includes('flit')) {
    build = 'flit';
  }

  const requiresPython = typeof project['requires-python'] === 'string' ? project['requires-python'] : '';
  const minVersion = requiresPython.replace(/^[\s>=^~]+/, '');

  const stack: TechStack = {runtime: 'python', language: 'python'};
  if (testRunner) {
    stack.test = testRunner;
  }
  if (build) {
    stack.build = build;
  }
  if (minVersion) {
    stack.min_version = minVersion;
  }
  return stack;
}

function detectRust(root: string): TechStack {
  let data: unknown;
  try {
    data = tomlParse(readFileSync(join(root, 'Cargo.toml'), 'utf-8'));
  } catch {
    return {};
  }
  if (!isPlainObject(data)) {
    return {};
  }
  const pkg = isPlainObject(data['package']) ? (data['package'] as Record<string, unknown>) : {};
  const rustVersion = typeof pkg['rust-version'] === 'string' ? pkg['rust-version'] : '';
  const stack: TechStack = {runtime: 'rust', language: 'rust', test: 'cargo', build: 'cargo'};
  if (rustVersion.length > 0) {
    stack.min_version = rustVersion;
  }
  return stack;
}

function detectGo(root: string): TechStack {
  let text: string;
  try {
    text = readFileSync(join(root, 'go.mod'), 'utf-8');
  } catch {
    return {};
  }
  const match = /^go\s+(\S+)/m.exec(text);
  const minVersion = match !== null ? match[1]! : '';
  const stack: TechStack = {runtime: 'go', language: 'go', test: 'go', build: 'go'};
  if (minVersion) {
    stack.min_version = minVersion;
  }
  return stack;
}
